### ISSUE TITLE
Enhance timestamp rounding algorithm

### ISSUE THREAD

**Issue by beevee** [moira-alert/moira#514]
Consider a metric that arrives _approximately_ every 60 seconds:

Original Timestamp | Rounded by Moira | Rounded by Graphite
------------------- | ------------------ | ---------------------
00:29                       | 00:00                      | 00:00
01:29                       | 01:00                       | 01:00
02:30                       | 03:00                      | 02:00
03:29                       | 03:00 (last write wins) | 03:00

From Graphite's standpoint, everything is fine: a point every minute. From Moira's standpoint, this metric has holes, and it can lead to unexpected alerts.

Why not just switch to the Graphite rounding algoritm? Consider a metric that arrives _approximately_ every 60 seconds:

Original Timestamp | Rounded by Moira | Rounded by Graphite
------------------- | ------------------ | ---------------------
00:00                       | 00:00                      | 00:00
01:00                       | 01:00                       | 01:00
01:59                       | 02:00                      | 01:00 (last write wins)
03:00                       | 03:00                      | 03:00

This time, everything is fine from Moira's standpoint. But there will be a hole in the Graphite graph.

A better rounding algorithm proposal for Moira:

1. First, try rounding the timestamp down, like Graphite.
2. If rounded-down slot is already occupied by a point, consider rounding up. Round up only if the original timestamp is close to the rounded-up timestamp. Let's say no more than 10% of retention away.
2. If rounding up is not allowed, overwrite rounded-down slot. Last write wins.

Original Timestamp | Rounded by Moira (proposed algorithm) | Rounded by Graphite
------------------- | ------------------ | ---------------------
00:29                       | 00:00                      | 00:00
01:29                       | 01:00                       | 01:00
02:30                       | 02:00                      | 02:00
03:29                       | 03:00                      | 03:00

Original Timestamp | Rounded by Moira (proposed algorithm) | Rounded by Graphite
------------------- | ------------------ | ---------------------
00:00                       | 00:00                      | 00:00
01:00                       | 01:00                       | 01:00
01:59                       | 02:00                      | 01:00 (last write wins)
03:00                       | 03:00                      | 03:00

**Comment by beevee** [comment]
Interesting places in code:

1. Rounding: https://github.com/moira-alert/moira/blob/a19b756b9ee9ef4f5ac3547e88e1dae688a2bb81/helpers.go#L195-L197
2. Also rounding: https://github.com/moira-alert/moira/blob/a19b756b9ee9ef4f5ac3547e88e1dae688a2bb81/metric_source/local/fetchdata.go#L94-L99
2. Data deduplication: https://github.com/moira-alert/moira/blob/a19b756b9ee9ef4f5ac3547e88e1dae688a2bb81/metric_source/local/fetchdata.go#L67-L69

**Comment by beevee** [comment]
An interesting case, when user erroneously sends points every 10 seconds instead of every minute. Let's say we allow rounding up within 10 seconds of the next timestamp.

Original Timestamp | Rounded by Moira (proposed algorithm)
------------------- | ------------------
00:10                       | 00:00
00:20                       | 00:00 (last write wins)
00:29                       | 00:00 (last write wins)
00:40                       | 00:00 (last write wins)
00:51                       | 01:00 (rounded up)
01:00                       | 01:00 (last write wins)
01:09                       | 01:00 (last write wins)
01:21                       | 01:00 (last write wins)
01:30                       | 01:00 (last write wins)

**Comment by hv7214** [comment]
@beevee  what i understood is : 
1. `roundToMinimalHighestRetention` is a kind of ceil of function.
2. `RoundToNearestRetention` is kind of floor function.

So does it meant to be merged to give it shape of proposed algorithm?

**Comment by beevee** [comment]
> RoundToNearestRetention is kind of floor function

No, it isn't. "Round to nearest" is not the same as "round down".

And I'm not sure we can merge these functions. For some reason they behave differently, so we should first understand why. IIRC, @borovskyav told me that he knows why these rounding functions are different.

**Comment by hv7214** [comment]
> For some reason they behave differently

Yes they do because `roundToMinimalHighestRetention` tries to round by adding full retention and latter tries to round up by adding half retention which in turn result in floor behaviour. 

For an instance: 
Let's say retention Is 60secs and time is 00:01. 
roundToMinimalHighestRetention returns : 01:00 (1 + 60 / 60 * 60)
RoundToNearestRetention returns : 00:00 (1 + 30 / 60 * 60)

**Comment by beevee** [comment]
I know they behave differently. Question is: why? If we mindlessly change both functions without trying to think why they were different in the first place, we risk breaking something.

**Comment by hv7214** [comment]
> Consider a metric that arrives _approximately_ every 60 seconds:
> 
> Original Timestamp	Rounded by Moira	Rounded by Graphite
> 00:29	00:00	00:00
> 01:29	01:00	01:00
> 02:30	03:00	02:00
> 03:29	03:00 (last write wins)	03:00

`RoundToNearestRetention` is responsible for this. 
In `plotting.go` above function is being used for rounding which in turn generates holes.
https://github.com/moira-alert/moira/blob/ea9670976c789e68b7c2454662cb7cf503f54021/notifier/plotting.go#L107-L109.
